envoyauth/response.go

```diff
@@ -343,54 +331,112 @@ func makeHeaderValueOption(k, v string) *ext_core_v3.HeaderValueOption {
 	}
 }
 
-func makeEnvoyHeaderValueOptionsFromHeadersMap(hvo []*ext_core_v3.HeaderValueOption, headers map[string]any) ([]*ext_core_v3.HeaderValueOption, error) {
-	hvo = slices.Grow(hvo, len(headers))
+func (result *EvalResult) getHeadersFromDecision(fieldName string) ([]*ext_core_v3.HeaderValueOption, error) {
+	return getHeadersWithTransformation(result.Decision, fieldName, preallocateForEnvoyHeaders, collectEnvoyHeaders)
+}
+
+func (result *EvalResult) getHTTPHeadersFromDecision(fieldName string) (http.Header, error) {
+	return getHeadersWithTransformation(result.Decision, fieldName, preallocateForHTTPHeaders, collectHTTPHeaders)
+}
+
+func getHeadersWithTransformation[T any](
+	decision any,
+	fieldName string,
+	preallocate func(*T, int),
+	collector func(string, string, *T),
+) (T, error) {
+	var result T
+
+	switch decision := decision.(type) {
+	case bool:
+		return result, nil
+	case map[string]interface{}:
+		headersList, err := extractHeadersFromDecision(decision, fieldName)
+		if err != nil {
+			return result, err
+		}
+
+		for _, headers := range headersList {
+			if err := collectHeaderValues(headers, preallocate, collector, &result); err != nil {
+				return result, err
+			}
+		}
+		return result, nil
+	default:
+		return result, fmt.Errorf("illegal value for policy evaluation result: %T", decision)
+	}
+}
+
+func extractHeadersFromDecision(decision map[string]interface{}, fieldName string) ([]map[string]interface{}, error) {
+	val, ok := decision[fieldName]
+	if !ok {
+		return nil, nil
+	}
+
+	switch v := val.(type) {
+	case map[string]interface{}:
+		return []map[string]interface{}{v}, nil
```

Comment:
Please, avoid additional allocations. We have a highload service based on OPA where each allocation drastically impacts the performance.

Reply:
```suggestion
func (result *EvalResult) getHTTPHeadersFromDecision(fieldName string) (http.Header, error) {
	opts, err := result.getHeadersFromDecision(fieldName)
	if err != nil {
		return nil, err
	}
	header := make(http.Header, len(opts))
	for _, opt = range opts {
		header.Add(opt.GetHeader().GetKey(), opt.GetHeader().GetValue())
	}
	return header, nil
}
```